### PR TITLE
518 fix AP Scorer

### LINKED ISSUE #518: AP@10 computation is incorrect
**Describe the bug**
AP@10 is not computed according to the formulation of one precision point for each relevant document.

**To Reproduce**
In Quepid, using AP@10 as the metric, search for Princess Bride in tmdb
The top ranked document is Princess Bride. Rate it 1 (relevant)
rate the remaining nine documents 0 (irrelevant)
AP@10 is shown as 0.29

**Expected behavior**
AP@10 should be 1.0

**Screenshots**
Screen Shot 2022-05-24 at 3 01 43 PM

**Additional context**

When computing average precision, we compute a precision value each time we encounter a relevant document in the ranked list. We then divide the sum of those precision points by the total number of relevant documents. Doing so allows us to treat all missing relevant documents as occurring at rank infinity, effective precision 0. This is discussed in [https://stackoverflow.com/questions/49275764/ir-calculate-average-precision-from-varying-relevant-documents-to-rank-k](url)

trec_eval [https://github.com/usnistgov/trec_eval](url) produces the expected behavior:

dfisher@earthsea test % cat res.ten
301	Q0	FR940202-2-00150	1	  2.129133		STANDARD
301	Q0	FR940202-2-00151	2	  1.724760		STANDARD
301	Q0	FR940202-2-00154	3	   2.104024	        STANDARD
301	Q0	FR940203-1-00036	4	  1.912871		STANDARD
301	Q0	FR940203-1-00038	5	  1.800881		STANDARD
301	Q0	FR940203-1-00039	6	  1.870957		STANDARD
301	Q0	FR940216-1-00014	7	  1.677013		STANDARD
301	Q0	FR940303-1-00006	8	  1.999081		STANDARD
301	Q0	FR940303-1-00012	9	  1.841413		STANDARD
301	Q0	FR940303-1-00014	10	  1.732111		        STANDARD
dfisher@earthsea test % cat qrel.ten
301	Q0	FR940202-2-00150	1
301	Q0	FR940202-2-00151	0
301	Q0	FR940202-2-00154	0
301	Q0	FR940203-1-00036	0
301	Q0	FR940203-1-00038	0
301	Q0	FR940203-1-00039	0
301	Q0	FR940216-1-00014	0
301	Q0	FR940303-1-00006	0
301	Q0	FR940303-1-00012	0
301	Q0	FR940303-1-00014	0
dfisher@earthsea test % ../trec_eval qrel.ten res.ten
runid                 	all	STANDARD
num_q                 	all	1
num_ret               	all	10
num_rel               	all	1
num_rel_ret           	all	1
map                   	all	1.0000
gm_map                	all	1.0000
…
P_5                   	all	0.2000
P_10                  	all	0.1000
P_15                  	all	0.0667
P_20                  	all	0.0500
P_30                  	all	0.0333
P_100                 	all	0.0100
P_200                 	all	0.0050
P_500                 	all	0.0020
P_1000                	all	0.0010

### PR DESCRIPTION
**Pull request by david-fisher**
## Description

Fix computation of AP to match trec_eval.
Update the ScorerFactory default scorer to use the new implementation.

## Motivation and Context

AP is the sum of the precisions at each rank where a relevant document occurs, divided by the total number of relevant documents (any not in the ranked list are treated as occurring at rank infinity with a precision of 0).

closes #518 

## How Has This Been Tested?

Manually verified AP computation in Quepid compared to the output of trec_eval.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [] Improvement (non-breaking change which improves existing functionality)
- [] New feature (non-breaking change which adds new functionality)
- [] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by epugh** [comment]
@worleydl @nathancday would love your input on this as you guys have poked a bit at AP as well....

**Comment by nathancday** [comment]
I like this PR for the trec_eval AP, thanks David, Ileft some code nits. I think TREC version should be the default AP

I'd like to keep the legacy implementation though for side by side, we can name it something else, maybe `ap_rolling`. IMO the TREC AP version is more useful when you can assume all of the relevant docs labeled, which TREC does via pooling, but I don't think we can assume always in our minimal test collections.

File: db/scorers/ap@10.js [hunk 1]
```diff
@@ -1,12 +1,25 @@
...
 eachDoc(function(doc, i) {
-    total += avgRating(i+1)
+    if (hasDocRating(i) && (docRating(i)) > 0) {
+        count++;
+    total += count/(i+1)
+    }
 }, k);
-
-var score = total / k;
+// count up the total number of relevant (not judged) documents
```
[the remark this answers is not in the document]
Comment (david-fisher) [reply db/scorers/ap@10.js:15]: The set of judged documents include those that have been judged irrelevant. The divisor is |R|, where R is the set of all documents d such that judgment(d) > 0.

File: app/assets/javascripts/factories/ScorerFactory.js [hunk 2]
```diff
@@ -11,16 +11,30 @@
...
+	  'let totalRel = 0;',
+	  'total = 0;',
+	  '// if less than K results, need to reduce K now or final score is too low',
+	  'k = numReturned() < k ? numReturned() : k',
+	  '// for each returned document, calculate precision each time a new',
+	  '// relevant document is added to the ranked list.',
+	  'eachDoc(function(doc, i) {',
+	  'if (hasDocRating(i) && (docRating(i)) > 0) {',
+          'count++;',
+	  'total += count/(i+1)',
```
[the remark this answers is not in the document]
Comment (david-fisher) [reply app/assets/javascripts/factories/ScorerFactory.js:26]: No. avgRating, which uses baseAvg, sums up the judgment values. AP computed on multivalued relevance judgments converts those judgments to binary values.  Using avgRating would not compute AP.

File: db/scorers/ap@10.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathancday) [reply db/scorers/ap@10.js:15]: I think the comment is confusing for me, how is a doc known to be relevant if it is not judged?

File: app/assets/javascripts/factories/ScorerFactory.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (nathancday) [reply app/assets/javascripts/factories/ScorerFactory.js:26]: good point I was assuming it would always be a binary scale, which would be the same as P@k

**Comment by nathancday** [comment]
My goal with the rolling-AP idea was to have an inherent positional discount for binary rating scales. I wrote this code before I learned about trec-eval and incorrectly assumed average precision was the average of precision values leading up to @K `sum(p@1, ..., p@k) / k`, not just the precise ones. I think there is similar utility in the output values.

I understood the goal of pooling to be about capturing all relevant documents, by combining diverse rankers with the assumption that anything not capture by any ranker is not relevant. I'm skeptical about getting this exhaustive level of judgments on minimal collections for clients. I'm also skeptical about assuming unjudged docs as irrelevant in these minimal collection, there are various methods published in that area. But I agree most IR metrics assume exhaustive recall.

Thanks for bringing Quepid to line up with TREC for AP. It looks great, looking forward to learning with you.

File: db/scorers/ap@10.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (david-fisher) [reply db/scorers/ap@10.js:15]: A doc in the set of all judged documents may have a score of 0, which is not relevant. This is strictly an issue with the data structure in Quepid.

[the remark this answers is not in the document]
Comment (epugh) [reply db/scorers/ap@10.js:15]: is this something else that we need to fix then?

[the remark this answers is not in the document]
Comment (david-fisher) [reply db/scorers/ap@10.js:15]: No. Counting how many relevant documents there are is only necessary when computing AP or recall. It is not expensive. Having the documents with judgments of 0 in the set of all judged documents is just fine (and probably used in other parts of the UI).

**Review by epugh — APPROVED** [review]

**Pull request merged**